drivers/isl/isl.cpp

```diff
@@ -0,0 +1,350 @@
+  write_reg(0x90, 0x0F); //Driver Range
+  write_reg(0x91, 0x80); //Emitter DAC
+  write_reg(0x92, 0x00); //Driver Control
+  write_reg(0x93, 0x00); //Thresh DAC
+  write_reg(0x94, 0x00); //Driver Boost
+  write_reg(0x96, 0x00); //Driv Chg Bal DAC
+  write_reg(0x97, 0x00); //Front End Ctl
+  write_reg(0x98, 0x00); //AFE Ctl Regs
+  write_reg(0xA5, 0x00); //Emitter ADC Offset
+  write_reg(0xA7, 0x00); //Temp Sens Reg A
+  write_reg(0xA8, 0x00); //Temp Sens Reg B
+  write_reg(0xA9, 0x00); //Temp Sens Reg C
+  write_reg(0xAA, 0x00); //Temp Sens ADC Mode
+  write_reg(0xae, 0x00); //Internal Reset
+  write_reg(0x31, 0x00); //Temp Reference
+  write_reg(0x33, 0x00); //Phase Exponent
+  write_reg(0x34, 0x00); //Temp Coeff 1
+  write_reg(0x39, 0x00); //Temp Coeff 2
+  write_reg(0x36, 0x00); //Ambient Coeff 1
+  write_reg(0x3B, 0x00); //Ambient Coeff 2
+  unsigned char sample_delay_reg = read_reg(0x10); // Sample Len
+  _isl_sample_delay = calc_sample_delay(sample_delay_reg);
+  return 0;
+}
+
+//Calculate sample period as per value received from Sensor
+//calculations based on information provided by intersil
+int ISL::calc_sample_delay(unsigned char value)
+{
+  float d = 0.0711 * pow(2, value);
+  return (int) d + 1;
+}
+
+//Register write function call for I2C bus
+int ISL::write_reg(uint8_t address, uint8_t data)
+{
+	return _writeReg(address, &data, sizeof(data));
+}
+
+//Register read function call for I2C bus
+uint8_t ISL::read_reg(uint8_t address)
+{
+	uint8_t ret = 0;
+	if(_readReg(address, &ret, sizeof(ret)) == -1) {
+		_read_failure = true;
+		DF_LOG_ERR("Read Failure");
+	}
+	return ret;
+}
+
+//Stop ISL distance sensor driver
+int ISL::stop()
+{
+	int result = DevObj::stop();
+
+	if (result != 0) {
+		DF_LOG_ERR("DevObj stop failed");
+		return result;
+	}
+
+	usleep(100000);
+
+	result = devClose();
+
+	if (result != 0) {
+		DF_LOG_ERR("device close failed");
+		return result;
+	}
+
+	usleep(100000);
+	return 0;
+}
+
+//ISL sensor calibration routine:
+//User is required to put sensor at REF_DIST and run this routine to get
+//calibration value.
+//TODO: Make the calculated offset persistent using px4 param system
+int ISL::calibration(void)
+{
+	uint16_t raw_data;
+	float raw_avg = 0;
+	write_reg(0x2f, 0);	//DIST_CAL_OFF_MSB
+	write_reg(0x30, 0);	//DIST_CAL_OFF_LSB
+	usleep(100000);
+	for(uint8_t i = 0; i< 100; i++) {
+		raw_data = read_reg(0xd2);		//DIST_REG_LSB
+		raw_data |= (read_reg(0xd1))<<8; //DIST_REG_MSB
+		raw_avg += float(raw_data)/100;
+	}
+	raw_avg /= 2000;	//convert to meters
+	float offset;
+	offset = (raw_avg - REF_DIST)*2000;
+	if(offset < 0) {
+		return -1;
+	}
+	uint8_t off_lsb, off_msb;
+	off_lsb = (0xFF & uint16_t(offset));
+	off_msb = (0xFF00 & uint16_t(offset))>>8;
+
+	write_reg(0x2f, off_msb);
+	write_reg(0x30, off_lsb);
+	return 0;
+}
+
+//Routine to scan through I2C bus for ISL sensor
+int ISL::_detect()
+{
+	uint8_t dev_id;
+	int result;
+	for(uint8_t i = 0; i <= 3; i++) {
```

Comment:
Try to not use <= as exit condition unless there is a real logic behind it. use < 4 instead.